src/XtensaOptimize.cpp

```diff
@@ -0,0 +1,2331 @@
+                // {"halide_xtensa_pred_sub_i16", wild_i16x - select(wild_u1x, wild_i16x, wild_i16x)},
+                // {"halide_xtensa_pred_sub_i32", wild_i32x - select(wild_u1x, wild_i32x, wild_i32x)},
+                {"halide_xtensa_widen_mul_sub_u24", wild_i24x - halide_xtensa_widen_mul_u24(wild_u8x, wild_u8x)},
+            };
+
+            Expr new_expr = apply_patterns(op, subs, this);
+            if (!new_expr.same_as(op)) {
+                return new_expr;
+            }
+        }
+
+        return IRGraphMutator::visit(op);
+    }
+
+    Expr visit(const Mul *op) override {
+        if (op->type.is_vector()) {
+            static const std::vector<Pattern> scalar_muls = {};
+
+            static const std::vector<Pattern> muls = {
+                {"halide_xtensa_widen_mul_i24", i24(wild_i8x) * bc(i24(wild_i8))},
+                {"halide_xtensa_widen_mul_u24", i24(wild_u8x) * bc(i24(wild_u8))},
+
+                {"halide_xtensa_widen_mul_i24", i24(wild_i8x) * i24(wild_i8x)},
+                {"halide_xtensa_widen_mul_u24", i24(wild_u8x) * i24(wild_u8x)},
+
+                {"halide_xtensa_widen_mul_by_diff_u24", (i24(wild_u8x) - bc(i24(wild_u8))) * bc(i24(wild_u8))},
+                {"halide_xtensa_widen_mul_by_diff_u24", (i24(wild_u8x) - bc(i24(wild_u8))) * i24(wild_u8x)},
+
+                {"halide_xtensa_widen_mul_i48", i48(wild_i16x) * i48(wild_i16x)},
+
+                {"halide_xtensa_mul_i32", wild_i32x * wild_i32x, Pattern::AccumulatorOutput64},
+
+                {"halide_xtensa_widen_zzzzz", i24(concat({wild_i8x64, wild_i8x64, wild_i8x64, wild_i8x64})) * i24(repeat_each_element(wild_i8x4, 64))},
+                {"halide_xtensa_widen_zzzzz", i24(wild_i8x256) * i24(repeat_each_element(wild_i8x4, 64))},
+                {"halide_xtensa_widen_zzzzz", i24(wild_u8x256) * bc(i24(wild_u8), 256)},
+                {"halide_xtensa_widen_zzzzz", i24(concat({wild_u8x64, wild_u8x64, wild_u8x64, wild_u8x64})) * i24(repeat_each_element(wild_u8x4, 64))},
+                {"halide_xtensa_widen_zzzzz", i24(wild_u8x256) * i24(repeat_each_element(wild_u8x4, 64))},
+
+                // Widening multiplication
+                // NOTE(vksnk): looked like a good idea, but seems to be slower. Need to double-check.
+                // {"halide_xtensa_widen_sqr_i48", wild_i32x * wild_i32x, Pattern::SameOp01 | Pattern::NarrowOps | Pattern::AccumulatorOutput48},
+            };
+
+            Expr new_expr = apply_commutative_patterns(op, scalar_muls, this);
+            if (!new_expr.same_as(op)) {
+                return new_expr;
+            }
+
+            new_expr = apply_commutative_patterns(op, muls, this);
+            if (!new_expr.same_as(op)) {
+                return new_expr;
+            }
+        }
+
+        return IRGraphMutator::visit(op);
+    }
+
+    Expr visit(const Div *op) override {
+        if (op->type.is_vector()) {
+            Expr div = op;
+            static const std::vector<Pattern> divs = {
```

Comment:
Are there ever cases where these rules trigger? `find_intrinsics` should convert divisions by powers of two into shifts already.